core-lib/demos/applications/InstantMessengerApplication.ns

```diff
@@ -0,0 +1,165 @@
+  private Dictionary = platform collections Dictionary.
+  private TransferObject = platform kernel TransferObject.
+|
+)(
+
+public class InstantMessenger new: name  total: size = (
+  | private buddyMap ::= Dictionary new: size.
+    private name = name.
+    private textMessage ::= nil.
+  |
+)(
+
+    public startChat: remoteMessenger = (
+      | pDiscover pName pSend msg pp array |
+
+        pp:: actors createPromisePair.
+
+        pDiscover:: self addMessenger: remoteMessenger. (* returns a far reference of the remote messenger *)
+
+        pName:: pDiscover <-: getName.
+        pName whenResolved: [:remoteName |
+                msg:: ('Hello ' + remoteName).
+                pSend:: sendMessage: remoteName contentMsg: msg.
+
+                pSend whenResolved:[:result |
+                    (result = 'ok')
+                    ifTrue:[
+                         array:: TransferArray new: 2. (* array with the msg to display in the sender actor screen *)
+                         array at: 1 put: name.
+                         array at: 2 put: msg.
+                         pp resolver resolve: array.
+                    ]
+                    ifFalse:[
+                         pp resolver resolve: nil.
+                    ]
+                ].
+        ].
+
+
+        ^ pp promise
+    )
+
+    public getName = (
+      ^ name
+    )
+
+    public sendMessage: receiverName contentMsg: content = (
+      | pp receiverActor pReceive|
+
+      textMessage:: TextMessage new: content sender: getName.
+
+      receiverActor:: buddyMap at: receiverName.
+      pReceive:: receiverActor <-: receive: textMessage.
+
+      ^ pReceive
+    )
+
+    public displayMessage: msg = (
+      (getName +' screen> '+ msg) println
+    )
+
+    public addMessenger: messenger = (
+      | p name pName buddy |
+
+         p:: actors createPromisePair.
+
+         pName:: messenger <-: getName.
+         pName whenResolved:[:n |
+           (buddyMap containsKey: n)
+           ifTrue:[
+             buddyMap at: n put: messenger. (* update el far reference *)
+             (getName + ' updated the far reference to buddy: '+ n) println.
+           ]
+           ifFalse:[
+             buddyMap at: n put: messenger.
+             (getName + ' discovered a new messenger buddy: '+ n) println.
+           ].
+           p resolver resolve: (buddyMap at: n).
+
+           ] onError:[:e|
+              ('-Error adding the messenger: ' + e) println.
+              p resolver resolve: nil.
+            ].
+
+      ^ p promise
+    )
+
+    public receive: textMessage = (
+      | sender content p |
+
+      content:: textMessage getContent.
+      sender:: textMessage getSender.
+
+      (* print message to the screen *)
+      self displayMessage:  sender +': '+ content.
+
+      (* acknowledgement notification of message received *)
+      ^ 'ok'.
+    )
+)
+
+ class TextMessage new: content sender: senderName = TransferObject (
+  | private msg = content.
```

Comment:
Since these slots are immutable, it seems unnecessary to have getters (`getContent`, `getSender`). Instead, these slots could simply be made `public`.

Reply:
fixed